### ISSUE TITLE
Microsoft.CodeAnalysis.CSharp.CommandLine.UnitTests.CommandLineTests.NoWarnAndWarnAsError_WarningDiagnostic failed in test runs.

### ISSUE THREAD

**Issue by balajikris** [dotnet/roslyn#11368]
Please diagnose if this test is flakey

Failed runs:
http://dotnet-ci.cloudapp.net/job/roslyn-internal_future_win_rel_unit32/385/
http://dotnet-ci.cloudapp.net/job/roslyn_future_win_rel_unit64/265/
http://dotnet-ci.cloudapp.net/job/roslyn-internal_future_win_dbg_unit32/390/
http://dotnet-ci.cloudapp.net/job/roslyn-internal_future_win_dbg_unit64/386/

~~~
MESSAGE:
Assert.Equal() Failure\r\nExpected: 2\r\nActual: 1
+++++++++++++++++++
STACK TRACE:
at Microsoft.CodeAnalysis.CSharp.CommandLine.UnitTests.CommandLineTests.VerifyOutput(TempDirectory sourceDir, TempFile sourceFile, Boolean includeCurrentAssemblyAsAnalyzerReference, String[] additionalFlags, Int32 expectedInfoCount, Int32 expectedWarningCount, Int32 expectedErrorCount) in q:\roslyn\src\Compilers\CSharp\Test\CommandLine\CommandLineTests.cs:line 7455 
at Microsoft.CodeAnalysis.CSharp.CommandLine.UnitTests.CommandLineTests.NoWarnAndWarnAsError_WarningDiagnostic() in q:\roslyn\src\Compilers\CSharp\Test\CommandLine\CommandLineTests.cs:line 7775
~~~

**Comment by balajikris** [comment]
This is our top failing issue now and has showed up atleast 4 times today, so far. Please help investigating this.

cc: @jaredpar

**Comment by jaredpar** [comment]
@VSadov please take a look at this.  

Note: this is an urgency now issue as it's causing lab instability.  Please give it the highest priority.

**Comment by VSadov** [comment]
Did it ever happen in Master or it is a Future only?

**Comment by balajikris** [comment]
you can mine that information from @jaredpar 's dashboard.

**Comment by jaredpar** [comment]
Here are all the failures since Firday including PRs

http://jdash.azurewebsites.net/builds/test?name=Microsoft.CodeAnalysis.CSharp.CommandLine.UnitTests.CommandLineTests.NoWarnAndWarnAsError_WarningDiagnostic&pr=true&startDate=2016-05-14

**Comment by jasonmalinowski** [comment]
This also failed in last night's signed build.

**Comment by VSadov** [comment]
The root cause is this (in CommonCompiler.cs):
```C#
             // GetDiagnosticsAsync is called after ReportUnusedImports
             // since that method calls EventQueue.TryComplete. Without
             // TryComplete, we may miss diagnostics.
             var hostDiagnostics = analyzerDriver.GetDiagnosticsAsync(compilation).Result;
```

The problem is that ReportUnusedImports above has a lot of conditions and is not always called.
Since GetDiagnosticsAsync simply grabs what is already in the queue, if compiling happens to not be finished, it will miss diagnostics.

Now need to figure how to wait for compiling to finish without deadlocking.

**Comment by jaredpar** [comment]
@VSadov can we disable this suite while we continue the investigation?  It's causing a number of false failures in the lab. 

@jcouv

**Comment by VSadov** [comment]
It seems while there is a way to guarantee that we are done compiling, there is no easy way to guarantee that analyzers are done analyzing and reporting diagnostics.

If we do not wait for analyzers, we can miss diagnostics.

In https://github.com/dotnet/roslyn/pull/11426 I have added intentional delays in the codepath that stores items into a diagnostic queue, and now it is much easier to hit the problem where analyzer diagnostics gets lost.

So, basically, I found easier way to reproduce, but not sure how to fix yet.
